### ISSUE TITLE
Maybe no-unused-vars should treat destructing differently

### ISSUE THREAD

**Issue by NickHeiner** [eslint/eslint#10543]
In an argument list, an argument is not considered "unused" if an argument after it is:

```js
// No lint errors
fn(unusedButWillNotCauseError, used) {
  console.log(used);
}
```

I think that perhaps list destructuring in an argument list should be treated the same way.

**What rule do you want to change?**
`no-unused-vars`

**Does this change cause the rule to produce more or fewer warnings?**
Fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**
Long term, I'd argue that the default behavior should be changed. Short term, for compatibility, it's probably best to use an option.

**Please provide some example code that this change will affect:**

```js
    foo([unused, used]) {
      console.log(used);
    }
```

**What does the rule currently do for this code?**
The rule flags `unused` as an unused variable.

**What will the rule do after it's changed?**
I suggest that maybe the rule should not flag `unused` as an unused variable, for the same reason that the following code would not cause an error:

```js
fn(unusedButWillNotCauseError, used) {
  console.log(used);
}
```

**Comment by platinumazure** [comment]
Array destructuring is different from argument lists in one crucial way: It is impossible to skip an argument (while still ensuring later arguments are in the right place); but you absolutely can skip an element in array destructuring:

```js
    foo([, used]) {
      console.log(used);
    }
```

So in that regard, I think the rule is working as intended.

**Comment by NickHeiner** [comment]
![mind blown](https://user-images.githubusercontent.com/829827/42117899-571251bc-7bb4-11e8-9bd1-8860fae0af8c.gif)

I had no idea you could do that. Ok, then this rule is working as intended. Thanks! (And thanks particularly for the lightning-fast response.)
